micronota/workflow.py

```diff
@@ -0,0 +1,152 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2015--, micronota development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from os.path import splitext, basename, join, exists
+from os import remove, makedirs
+from importlib import import_module
+from tempfile import NamedTemporaryFile
+
+from skbio.metadata import IntervalMetadata
+from skbio import read, Sequence
+
+from . import bfillings
+from .util import _DB_PATH
+
+
+def annotate(in_fp, in_fmt, out_dir, out_fmt, kingdom, cpus, config):
+    '''Annotate the sequences in the input file.'''
+    fn = splitext(basename(in_fp))[0]
+    # store annotated seq file.
+    makedirs(out_dir, exist_ok=True)
+    out_fp = join(out_dir, '%s.gbk' % fn)
+    out = open(out_fp, 'w')
+    for seq in read(in_fp, format=in_fmt):
+        # dir for useful intermediate files for the current input seq
+        seq_tmp_dir = join(out_dir, seq.metadata['id'])
+        for sec in config:
+            if sec == 'GENERAL':
+                continue
+            if sec == 'FEATURE':
+                im = identify_all_features(seq, seq_tmp_dir, config[sec])
+                seq.interval_metadata = IntervalMetadata(im)
+            if sec == 'CDS':
+                im = annotate_all_cds(seq, out_dir, kingdom, config[sec])
+
+        seq.interval_metadata.concat(im, inplace=True)
+        seq.write(out, format=out_fmt)
+    out.close()
+
+
+def identify_all_features(seq, out_dir, tasks,
+                          id_func='identify_features',
+                          parse_func='parse_output'):
+    '''Identify all the features for the sequence in the input file.
+
+    It runs thru all the tasks specified in sequential order.
+
+    Parameters
+    ----------
+    tasks : OrderDict-like
+    Returns
+    -------
+    '''
+    im = dict()
+    with NamedTemporaryFile('w+') as f:
+        seq.write(f.name, format='fasta')
+
+        for tool in tasks:
+            try:
+                value = tasks.getboolean(tool)
+            except ValueError:
+                value = tasks[tool]
+            if not value:
+                continue
```

Comment:
@RNAer , what exactly is going on here?   What type of object is tasks?  What is value?